### DISCUSSION TITLE
Getting transformation map of ElasticDeformation

### DISCUSSION THREAD

**Discussion by timroelofs123** [TorchIO-project/torchio#524] [Q&A] [answered]
Hi, I was wondering if there exists a method to get a displacement vector field of the (Random)ElasticDeformation function, i.e. an array where the displacement of every voxel is given in three directions. 

Getting the displacement at the control points is simple, but I am unsure how I can get from this to the displacement at all voxel locations, i.e. how the interpolation works - my trials to reproduce it using linear interpolation in scipy did not work. 

The reason I am asking is not to reapply the transformation, but because I am working on a registration model, and I wanted to see if the model could predict the displacement of each voxel applied with the RandomElasticDeformation function. 

My method right now is to pass an array where every voxel has a unique index value to the function, and derive it from this - but while this does give the DVF (rounded to integer values) it is not every efficient. 

Any help would be appreciated!

**Comment by fepegar** [comment, marked as answer]
Hi, @timroelofs123. Maybe you can get the transform as I explained in [this comment](https://github.com/fepegar/torchio/issues/442#issuecomment-773210906) and then use [sitk.TransformToDisplacementFieldFilter](https://simpleitk.org/doxygen/latest/html/classitk_1_1simple_1_1TransformToDisplacementFieldFilter.html#details) to create the field. Please let me know if that works.

**Reply by timroelofs123** [reply]
Thanks for your help! I am not well versed in SITK but that indeed did the trick. 
For future reference in case anyone else wants to do the same thing, here is my function: 

```
def get_deformation_grid(subject): 
    '''
    param subject:  a transformed subject containing a ScalarImage 'scan'
    return: displacement field of shape [1, W, L, H, 3] with the last dimension being the deformation in [x,y,z].
    '''
    deterministic = subject.history[-1]
    itk_transform = deterministic.get_bspline_transform(subject.scan.as_sitk(), control_points = deterministic.control_points)
    transform_to_displacement_field_filter = sitk.TransformToDisplacementFieldFilter()
    transform_to_displacement_field_filter.SetReferenceImage(subject.scan.as_sitk())
    disp_transform = sitk.DisplacementFieldTransform(transform_to_displacement_field_filter.Execute(itk_transform))
    deformation_map = sitk.GetArrayFromImage(disp_transform.GetDisplacementField())
    deformation_map = np.transpose(deformation_map, axes=[2, 1, 0, 3])[np.newaxis, ...]*(-1,-1,1) # Correct shape for my purpose
    return deformation_map
```

**Reply by fepegar** [reply]
I'm glad it worked. Thanks for sharing your solution!
